packages/firestore/test/integration/api/database.test.ts

```diff
@@ -542,6 +543,364 @@ apiDescribe('Database', (persistence: boolean) => {
     });
   });
 
+  apiDescribe('onSnapshotsInSync listener', () => {
```

Comment:
So I think Gil is write that these could be written as spec tests.  But this will require some plumbing work... basically the ability for spec tests to add / remove snapshots-in-sync listeners, and the ability for tests to express when the event is expected to be raised.  As a rough outline (probably not 100% right and all the details are TBD) I expect that you'll need to:

1. Introduce a new SpecBuilder method that we can call from spec tests similar to the existing userListen one: https://github.com/firebase/firebase-js-sdk/blob/beb72676a75dabf5e0f965f5ee06208b60c9ea45/packages/firestore/test/unit/specs/spec_builder.ts#L217
2. Introduce a new `SpecStep` type: https://github.com/firebase/firebase-js-sdk/blob/beb72676a75dabf5e0f965f5ee06208b60c9ea45/packages/firestore/src/core/event_manager.ts#L214
3. Make the `SpecTestRunner` code handle the new SpecStep: https://github.com/firebase/firebase-js-sdk/blob/beb72676a75dabf5e0f965f5ee06208b60c9ea45/packages/firestore/test/unit/specs/spec_builder.ts#L217
4. Need to add a way for tests to express expectations regarding when the event fires, similar to expectEvents(): https://github.com/firebase/firebase-js-sdk/blob/beb72676a75dabf5e0f965f5ee06208b60c9ea45/packages/firestore/test/unit/specs/spec_builder.ts#L722
5. Plumb the expected events into the `SpecStep` somehow.  This may mean adding a new field to `SpecStep` that's similar to `expect` (which really should be called `expectedEvents`) or else somehow extending the existing `SpecExpectation` so that it can cover both snapshot events and snapshots-in-sync events.  https://github.com/firebase/firebase-js-sdk/blob/beb72676a75dabf5e0f965f5ee06208b60c9ea45/packages/firestore/test/unit/specs/spec_test_runner.ts#L1406
6. Update our step runner code to validate the new expectations as appropriate. https://github.com/firebase/firebase-js-sdk/blob/beb72676a75dabf5e0f965f5ee06208b60c9ea45/packages/firestore/test/unit/specs/spec_test_runner.ts#L520

The benefit of all this is that you'll be able to write the tests once and then run them on all platforms (once you make the appropriate SpecTestRunner changes on the other platforms). They'll also be faster since they'll run as synchronous unit tests instead of as slow integration tests that are network-dependent.  And they'll hopefully end up more concise / more readable (less boilerplate), etc.  It's also a great opportunity for you to ramp up more deeply on spec tests! :-)

I would probably keep 1 or 2 integration tests that just smoke test the public API, but we can hopefully use spec tests to test all the edge cases.